src/Psalm/Internal/Analyzer/Statements/Expression/BinaryOp/ArithmeticOpAnalyzer.php

```diff
@@ -867,4 +893,134 @@ public static function arithmeticOperation(
 
         return $calculated_type;
     }
+
+    private static function analyzeOperandsBetweenIntRange(
+        PhpParser\Node $parent,
+        ?Type\Union &$result_type,
+        TIntRange $left_type_part,
+        TIntRange $right_type_part
+    ): void {
+        if ($parent instanceof PhpParser\Node\Expr\BinaryOp\Div) {
+            //can't assume an int range will stay int after division
+            if (!$result_type) {
+                $result_type = new Type\Union([new Type\Atomic\TInt(), new Type\Atomic\TFloat()]);
+            } else {
+                $result_type = Type::combineUnionTypes(
+                    new Type\Union([new Type\Atomic\TInt(), new Type\Atomic\TFloat()]),
+                    $result_type
+                );
+            }
+            return;
+        }
+
+        if ($parent instanceof PhpParser\Node\Expr\BinaryOp\BitwiseAnd ||
+            $parent instanceof PhpParser\Node\Expr\BinaryOp\BitwiseOr ||
+            $parent instanceof PhpParser\Node\Expr\BinaryOp\BitwiseXor
+        ) {
+            //really complex to calculate
```

Comment:
Intuitively, for positive ranges it's constrained by `int<0, max(a_max, b_max)>`, but this can be improved later.

Reply:
No, it's more complex than that unfortunately: https://3v4l.org/5bgJ9

At least for Or and Xor, the max can be higher than the max of both ranges. Worse, if you have `int<64,64>` and `int<62,64>` the max will be `63|64` so you have to go looking in between the bounds...

But yeah, I figured anyone who could benefit for a better inference here will have better knowledge than me in binary operations! Maybe there's a simple solution here I don't see.

Reply:
> At least for Or and Xor, the max can be higher than the max of both ranges

True. The result will, however, be of the same (binary) magnitude because none of those operations carry over (that's why they called bitwise). Basically we can throw off most significant `0` bits, fill the remaining with `1`s and call that max. That's for positive ints. Negatives are likely a different story, because they are stored as binary complements.

Reply:
Not sure with bits operation usage, but from what I've seen this far is that range operations can lead to false positive quickly (for example, using the range to hit an offset for a known array. If you don't end up with an exact answer, you risk something like `$array[$my_range]` where $my_range is off by a few integers but it's enough to create a false positive.